### PR TITLE
Update bounds

### LINKED ISSUE #63: Need to add info on how bounds on coordiantes are calculated
We should add some information on how the bounds on coordiantes are calculated, and how we use the IK_Bounds_Default.mot file

### LINKED ISSUE #85: Error when setting initial guess Qs
*Before submitting a bug report, read the error messages in MATLAB. Some errors are there to inform you about inconsistent inputs for certain settings.* 

**Describe the bug**
guess for coordinates mtj_angle_r and mtj_angle_l are NaN

**To Reproduce**
1. Use model with coordinates that are not included in IK_Bounds_default.mot

**Log file**
Error message:
> Error using casadi.Opti/set_initial (line 230)
Error in Opti::set_initial [OptiNode] at .../casadi/core/optistack.cpp:127:
.../casadi/core/optistack_internal.cpp:1149: Assertion "v.is_regular()" failed:
Notify the CasADi developers.
Error in OCP_formulation (line 143)
opti.set_initial(Qs, guess.Qs');
Error in run_pred_sim (line 127)
    OCP_formulation(S,model_info,f_casadi);
Error in main (line 136)
    [savename] = run_pred_sim(S,osim_path); 

**Software versions**
 - n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by Lars-DHondt-KUL**
Updates:
- Bounds on coordinates are stored as a table. This is more transparent than the calculations based on a motion file. Closes #63 The stored bounds are the same as used before. Added setting to select the table with default bounds.
- Added settings to specify bounds on coordinate velocities and accelerations
- If a coordinate bounds is not given by table or settings, it is computed from the range in the opensim model file. This avoids having upper and lower bound = 0, thus scale factor NaN. Closes #85
- Added settings to specify upper/lower bounds for activation of specific muscles. Can be useful to model increased baseline tone, or reduced maximum due to botox injections
- Added possibility to set scale factors for Qs, Qdots, Qdotdots, and for moment equilibrium constraints.

Function getBounds replaced by 3 functions:
- *getBounds* now only gets the bounds, they are unscaled
- *getScalefactor* computes the scale factors, based on the bounds just like before (but this opens up future explorations of decoupling bounds and scaling)
-  *scaleBounds* scales the bounds
The unscaled bounds are also saved in setup struct

Can replicate test results 'Falisse_et_al_2022'

### CODE REVIEW

File: DefaultSettings/getDefaultSettings.m [hunk 1]
```diff
@@ -65,22 +78,75 @@
...
+
+% minimal muscle activation, a number between 0 and 1
+if ~isfield(S.bounds.activation_all_muscles,'lower')
+    S.bounds.activation_all_muscles.lower = 0.05;
+end
+
+% activation of selected muscles
+if ~isfield(S.bounds,'activation_selected_muscles')
+    S.bounds.activation_selected_muscles = [];
+end
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:124]: I can see you can set the upper bound for activations using S.bounds.activation_selected_muscles, but would it be useful to be able to set the upper bounds of all muscles with just one setting (e.g. S.bounds.activation_all_muscles.upper)?

File: DefaultSettings/getDefaultSettings.m [hunk 2]
```diff
@@ -132,11 +201,21 @@
         num2str(S.misc.poly_order.lower) '_' num2str(S.misc.poly_order.upper)];
 end
 
+% default coordinate bounds used to approximate musculoskeletal geometry
+if ~isfield(S.misc,'default_msk_geom_bounds')
+    S.misc.default_msk_geom_bounds = 'default_msk_geom_bounds.csv';
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:206]: default_msk_geom_bounds.csv not present in this branch?

File: DefaultSettings/getDefaultSettings.m [hunk 3]
```diff
@@ -379,7 +477,20 @@
...
 
+% ligament stiffness for all ligaments
+if ~isfield(S.subject,'stiffness_all_ligaments')
+    S.subject.stiffness_all_ligaments = 'ligamentGefen2002'; 
+end
+
+% ligament stiffness for selected ligaments
+if ~isfield(S.subject,'set_stiffness_selected_ligaments')
+    S.subject.set_stiffness_selected_ligaments = {'PlantarFascia','plantarFasciaNatali2010'};
+end
```
Comment (VDB-Bram) [inline DefaultSettings/getDefaultSettings.m:488]: Is this related to the updated bounds?

File: OCP/AdaptGuess_UserInput.m [hunk 4]
```diff
@@ -28,7 +28,10 @@
 % Last edit date: 
 % --------------------------------------------------------------------------
 
-guess.a(guess.a < S.bounds.a.lower) = S.bounds.a.lower;
+for i=1:length(bounds.a.lower)
```
Comment (VDB-Bram) [inline OCP/AdaptGuess_UserInput.m:31]: For my understanding, why only do this for the activations? Wouldn't it be possible that your IG is also out of bounds when basing the bounds on a suggested Default_Coordinate_Bounds.csv file and not the IG itself like before?

File: OCP/AdaptGuess_UserInput.m [hunk 5]
```diff
@@ -28,7 +28,10 @@
...
+    guess.a(guess.a(:,i) < bounds.a.lower(i),i) = bounds.a.lower(i);
+end
+
```
Comment (VDB-Bram) [inline OCP/AdaptGuess_UserInput.m:34]: For my understanding, why only do this for the activations? Wouldn't it be possible that your IG is also out of bounds when basing the bounds on a suggested Default_Coordinate_Bounds.csv file and not the IG itself like before?

File: DefaultSettings/getDefaultSettings.m (hunk 1 above)
Comment (Lars-DHondt-KUL) [reply DefaultSettings/getDefaultSettings.m:124]: I have added a setting for the upper bound

File: DefaultSettings/getDefaultSettings.m (hunk 2 above)
Comment (Lars-DHondt-KUL) [reply DefaultSettings/getDefaultSettings.m:206]: looks like something went wrong with updating this branch from master, should be fixed now.

File: DefaultSettings/getDefaultSettings.m (hunk 3 above)
Comment (Lars-DHondt-KUL) [reply DefaultSettings/getDefaultSettings.m:488]: these settings also exist on master, so not sure why it shows up as a change here

File: OCP/AdaptGuess_UserInput.m (hunk 5 above)
Comment (Lars-DHondt-KUL) [reply OCP/AdaptGuess_UserInput.m:34]: ipopt will push the initial guess to fit between the bounds (https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_bound_push), so we don't really need to care about this.
I removed this function.

File: OCP/getBounds.m [hunk 6]
```diff
@@ -15,289 +11,175 @@
...
 
+%% Vertical position of floating base
+if ~isempty(S.bounds.factor_IG_pelvis_ty.lower)
+    bounds_nsc.Qs.lower(model_info.ExtFunIO.jointi.floating_base(5)) = model_info.IG_pelvis_y...
+        *S.bounds.factor_IG_pelvis_ty.lower;
+end
+if ~isempty(S.bounds.factor_IG_pelvis_ty.upper)
+    bounds_nsc.Qs.upper(model_info.ExtFunIO.jointi.floating_base(5)) = model_info.IG_pelvis_y...
+        *S.bounds.factor_IG_pelvis_ty.upper;
 end
```
Comment (VDB-Bram) [inline OCP/getBounds.m:122]: Is it so that in every model the vertical position of the floating base is the 5th coordinate?

File: OCP/getBounds.m [hunk 7]
```diff
@@ -15,289 +11,175 @@
...
+%% Final time
+bounds_nsc.tf.lower = S.bounds.t_final.lower;
+bounds_nsc.tf.upper = S.bounds.t_final.upper;
+
 
 %% Muscle-tendon forces
-bounds.FTtilde.lower = zeros(1,NMuscle);
-bounds.FTtilde.upper = 5*ones(1,NMuscle);
+bounds_nsc.FTtilde.lower = zeros(1,NMuscle);
+bounds_nsc.FTtilde.upper = 5*ones(1,NMuscle);
```
Comment (VDB-Bram) [inline OCP/getBounds.m:163]: For my insight; where does the 5 come from?

**Comment by VDB-Bram** [comment]
@Lars-DHondt-KUL , do you have a suggestion on how to best test this PR?

**Comment by VDB-Bram** [comment]
I ran main.m and got these results. Result seems to not be identical to the reference results:

1. Very small difference between arm kinematics
2. Very small difference in pelvis tilt
3. Very small difference in bifemlh_r activations at beginning and end
4. Different number of iterations

[update_bounds-results.zip](https://github.com/KULeuvenNeuromechanics/PredSim/files/14111182/update_bounds-results.zip)

image
image

**Comment by VDB-Bram** [comment]
Additional changes resolve the issues I had. I started the simulation and let it run for 20 iterations (see .txt attached). The objective is the same as the results above shared earlier. 


[Falisse_et_al_2022_v2_log.txt](https://github.com/KULeuvenNeuromechanics/PredSim/files/14181185/Falisse_et_al_2022_v2_log.txt)

**Review by VDB-Bram — APPROVED** [review]
## Do I understand the changes?

- [x] Functions and variables have meaningful names.
- [x] Functions have a header according to the [template](https://github.com/KULeuvenNeuromechanics/PredSim/wiki/Guidelines-for-contributing-code#use-this-template-for-functions).
- [x] The [readme](https://github.com/KULeuvenNeuromechanics/PredSim/blob/master/README.md) is up to date.

## Do the changes work?

- [x] The provided unit test return the expected result. Not all changes require a unit test. Reviewers can ask the contributor to provide a unit test if they think it is needed.
- [x] Simulations with settings as described in the pull request run without errors.

## Do the changes not break the existing code?

- [x] The changed code can reproduce simulation results for the example model and settings (i.e. [main.m](https://github.com/KULeuvenNeuromechanics/PredSim/blob/master/main.m) from the master branch).
